File: content/snippets/go/interview_lcci_17_13_re_space.go
package main

func respace(dictionary []string, sentence string) int {
	s := make(map[string]bool)
	for _, w := range dictionary {
		s[w] = true
	}
	n := len(sentence)
	dp := make([]int, n+1)
	for i := 1; i <= n; i++ {
		dp[i] = dp[i-1] + 1
		for j := 0; j < i; j++ {
			if s[sentence[j:i]] {
				if dp[j] < dp[i] {
					dp[i] = dp[j]
				}
			}
		}
	}
	return dp[n]
}

File: content/snippets/typescript/interview_lcci_17_13_re_space.ts
function respace(dictionary: string[], sentence: string): number {
    const s = new Set(dictionary);
    const n = sentence.length;
    const dp = new Array(n + 1).fill(0);
    for (let i = 1; i <= n; i++) {
        dp[i] = dp[i - 1] + 1;
        for (let j = 0; j < i; j++) {
            if (s.has(sentence.slice(j, i))) {
                dp[i] = Math.min(dp[i], dp[j]);
            }
        }
    }
    return dp[n];
}
